protocols/primary-backup/src/main/java/io/atomix/protocols/backup/service/impl/PrimaryBackupServiceContext.java

```diff
@@ -621,9 +621,13 @@ private void changeRole(PrimaryTerm term) {
    * Closes the service.
    */
   public CompletableFuture<Void> close() {
-    clusterMembershipService.removeListener(membershipEventListener);
-    primaryElection.removeListener(primaryElectionListener);
-    role.close();
-    return CompletableFuture.completedFuture(null);
+    CompletableFuture<Void> future = new CompletableFuture<>();
+    threadContext.execute(() -> {
+      clusterMembershipService.removeListener(membershipEventListener);
+      primaryElection.removeListener(primaryElectionListener);
+      role.close();
+      future.complete(null);
```

Comment:
I would consider wrapping completing the future in a finally block to ensure it's always set, imagine close of role is refactored and then throws an exception..

Reply:
true